packages/react/polymorphic/src/polymorphic.ts

```diff
@@ -41,15 +41,13 @@ interface ForwardRefComponent<
    * We explicitly avoid `React.ElementType` and manually narrow the prop types
    * so that events are typed when using JSX.IntrinsicElements.
    */
-  <
-    As extends
-      | keyof JSX.IntrinsicElements
-      | React.ComponentType<any> = NarrowIntrinsic<IntrinsicElementString>
-  >(
-    props: As extends keyof JSX.IntrinsicElements
-      ? Merge<JSX.IntrinsicElements[As], OwnProps & { as: As }>
+  <As = IntrinsicElementString>(
+    props: As extends ''
+      ? { as: keyof JSX.IntrinsicElements }
```

Comment:
Not super familiar with these typings. Is the gist that the conditionals are progressively narrowing from simple to more complex only as necessary?

Reply:
Yes, and we do this instead of checking `As extends React.ElementType` because that breaks event typing for intrinsics.